examples/research_projects/wav2vec2/alignment.py

```diff
@@ -0,0 +1,224 @@
+            """
+            num_frames = emission.size(0)
+            num_tokens = len(tokens)
+
+            # Trellis has extra diemsions for both time axis and tokens.
+            # The extra dim for tokens represents <SoS> (start-of-sentence)
+            # The extra dim for time axis is for simplification of the code.
+            trellis = torch.full((num_frames + 1, num_tokens + 1), -float("inf"))
+            trellis[:, 0] = 0
+            for t in range(num_frames):
+                trellis[t + 1, 1:] = torch.maximum(
+                    # Score for staying at the same token
+                    trellis[t, 1:] + emission[t, blank_id],
+                    # Score for changing to the next token
+                    trellis[t, :-1] + emission[t, tokens],
+                )
+            return trellis
+
+        trellis = get_trellis(emission, tokens, blank_id)
+
+        @dataclass
+        class Point:
+            token_index: int
+            time_index: int
+            score: float
+
+        def backtrack(trellis, emission, tokens, blank_id=0):
+            """
+            Walk backwards from the last (sentence_token, time_step) pair to build the optimal sequence alignment path
+            """
+            # Note:
+            # j and t are indices for trellis, which has extra dimensions
+            # for time and tokens at the beginning.
+            # When referring to time frame index `T` in trellis,
+            # the corresponding index in emission is `T-1`.
+            # Similarly, when referring to token index `J` in trellis,
+            # the corresponding index in transcript is `J-1`.
+            j = trellis.size(1) - 1
+            t_start = torch.argmax(trellis[:, j]).item()
+
+            path = []
+            for t in range(t_start, 0, -1):
+                # 1. Figure out if the current position was stay or change
+                # Note (again):
+                # `emission[J-1]` is the emission at time frame `J` of trellis dimension.
+                # Score for token staying the same from time frame J-1 to T.
+                stayed = trellis[t - 1, j] + emission[t - 1, blank_id]
+                # Score for token changing from C-1 at T-1 to J at T.
+                changed = trellis[t - 1, j - 1] + emission[t - 1, tokens[j - 1]]
+
+                # 2. Store the path with frame-wise probability.
+                prob = emission[t - 1, tokens[j - 1] if changed > stayed else 0].exp().item()
+                # Return token index and time index in non-trellis coordinate.
+                path.append(Point(j - 1, t - 1, prob))
+
+                # 3. Update the token
+                if changed > stayed:
+                    j -= 1
+                    if j == 0:
+                        break
+            else:
+                raise ValueError("Failed to align")
+            return path[::-1]
+
+        path = backtrack(trellis, emission, tokens)
```

Comment:
Should Pass in blank_id I believe

Reply:
Thanks @nabil6391 for noticing, I will validate my local code and update it here.